src/psfmachine/tpf.py

```diff
@@ -283,6 +294,130 @@ def plot_tpf(self, tdx):
+        except KeyError:
+            raise IOError(
+                "Quarter %i and channel %i has no PRF model data" % (quarter, channel)
+            )
+
+        # create source mask and uncontaminated pixel mask
+        self._get_source_mask()
+        self._get_uncontaminated_pixel_mask()
+
+        # create mean model, but PRF shapes from FFI are in pixels! and TPFMachine
+        # work in arcseconds
+        self._get_mean_model()
+
+        # Re-estimate source flux
+        # -----
+        prior_mu = self.source_flux_estimates
+        prior_sigma = (
+            np.ones(self.mean_model.shape[0]) * 10 * self.source_flux_estimates
+        )
+
+        f, fe = (self.flux).mean(axis=0), ((self.flux_err ** 2).sum(axis=0) ** 0.5) / (
+            self.nt
+        )
+
+        X = self.mean_model.copy()
+        X = X.T
+
+        sigma_w_inv = X.T.dot(X.multiply(1 / fe[:, None] ** 2)).toarray()
+        sigma_w_inv += np.diag(1 / (prior_sigma ** 2))
+        B = X.T.dot((f / fe ** 2))
+        B += prior_mu / (prior_sigma ** 2)
+        ws = np.linalg.solve(sigma_w_inv, B)
+        werrs = np.linalg.inv(sigma_w_inv).diagonal() ** 0.5
+
+        # -----
+
+        # Rebuild source mask
+        ok = np.abs(ws - self.source_flux_estimates) / werrs > 3
+        ok &= ((ws / self.source_flux_estimates) < 10) & (
+            (self.source_flux_estimates / ws) < 10
+        )
+        ok &= ws > 10
+        ok &= werrs > 0
+
+        self.source_flux_estimates[ok] = ws[ok]
+
+        self.source_mask = (
+            self.mean_model.multiply(
+                self.mean_model.T.dot(self.source_flux_estimates)
+            ).tocsr()
+            > 1
+        )
+
+        # Recreate uncontaminated mask
+        self._get_uncontaminated_pixel_mask()
+
+        # Recreate mean model!
+        self._get_mean_model()
+
+        if plot:
+            return self.plot_shape_model()
+        return
+
+    def save_shape_model(self, output=None):
+        """Saves the weights of a PRF fit to a file
+        Parameters
+        ----------
+        output : str, None
+            Output file name. If None, one will be generated.
+        """
+        # asign a file name
+        if output is None:
+            output = "%s/data/shape_model_ch%02i_q%02i.csv" % (
+                PACKAGEDIR,
+                self.tpf_meta["channel"][0],
+                self.tpf_meta["quarter"][0],
+            )
+
+        # create data structure (DataFrame) to save the model params
+        arr_to_save = np.array(
+            [self.n_r_knots, self.n_phi_knots, self.rmin, self.rmax]
+            + self.psf_w.tolist()
+        )
+        df_dict = {
+            self.tpf_meta["quarter"][0]: pd.DataFrame(
+                np.atleast_2d(arr_to_save),
+                index=[self.tpf_meta["channel"][0]],
+                columns=["n_r_knots", "n_phi_knots", "rmin", "rmax"]
+                + ["w%02i" % i for i in range(1, 1 + len(self.psf_w))],
+            )
```

Comment:
We should probably save some comment meta data about e.g. the spline order, and maybe the keywords that made the file so that it's reproducible?